walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,113 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.HashSet;
+
+/**
+* The FindMeetingQuery receives a new Meeting request to fit inside a
+* calendar day with existing meetings. It provides an interface to
+* handle incoming meetings that must be scheduled for users without
+* generating interference/overlapping with their existing meetings.
+* 
+* Algorithmic analysis
+* N = Scheduled events
+* M = Attendees within scheduled events
+*
+* Time complexity: N * M
+* Space complexity: N being the availableTimeRanges ArrayList since it is technically possible to have N available time ranges.
+*   This depends on the analysis rules that are set since all the linear structures
+*   that are used inside the method are for storing and parsing the incoming parameters.
+*   Thus, they shouldn't be considered as extra space because it's not space needed directly
+*   for the solution of the algorithm.
+*   The arrays minutesNoOptAttendees and minutesOptAttendees can be considered constant since
+*   they represent an abstraction of a whole day divided in minutes.
+*
+* @param Collection<Event> List of scheduled events.
+* @param MeetingRequest Event to fit in the available time windows.
+* @return Collection<TimeRange>
+*/
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        int TIME_WINDOW_IN_MINUTES = 60 * 24;
+        Integer[] minutesNoOptAttendees = new Integer[TIME_WINDOW_IN_MINUTES];
+        Integer[] minutesOptAttendees = new Integer[TIME_WINDOW_IN_MINUTES];
+
+        Arrays.fill(minutesNoOptAttendees, 1);
+        Arrays.fill(minutesOptAttendees, 1);
+
+        ArrayList<TimeRange> availableTimeRanges = new ArrayList<>();
+        ArrayList<Event> eventsToSet = new ArrayList<>(events);
+
+        Set<String> attendees = new HashSet<String>(request.getAttendees());
```

Comment:
This copy constructor is sneakily foiling your space complexity analysis. The `new HashSet<String>=(request.GetAttendees())` will allocate new memory. But you can just avoid this by doing `Set<String> attendees = request.getAttendees();` since I don't think you modify the contents anywhere.

Reply:
+1 to Annie's comment, with one additional suggestion: if you do not need to modify `attendees` in the subsequent code, I recommend using [`Collections.unmodifiableSet`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#unmodifiableSet-java.util.Set-) -- it will not copy the underlying set, but will cause an exception if any code tries to modify it.

Reply:
Sweet. Although, I will be switching from _Set_ to a _Collection_ structure to store both **attendees** and **optionalAttendees** since the `request.getAttendees()` return a Collection instead of a Set structure, and that was throwing me this error:

`incompatible types: java.util.Collection<java.lang.String> cannot be converted to java.util.Set<java.lang.String>`

But apart from that, it's working like a charm :)